### DISCUSSION TITLE
Prepared statement fails with more than 65,528 parameters

### DISCUSSION THREAD

**Discussion by bgrainger** [planetscale/discussion#371] [Bug Reports]
Executing a prepared statement with more than 65,528 parameters fails in two different ways:

## 65529 to 65535 parameters

When issuing `COM_STMT_EXECUTE` to execute the prepared statement, the server abruptly closes the connection.

## >= 65536 parameters

When executing the prepared statement, the server returns an ERR packet with a message like `target: mysqltest.-.primary: vttablet: rpc error: code = InvalidArgument desc = missing bind var v4465 (CallerID: xlc5dtqu00ynckn1otee)`.

## Expectation

If a client sends a `COM_STMT_PREPARE` packet with >= 65529 parameters, the server should return an ERR packet with error code `ER_PS_MANY_PARAM` (1390). It should not be possible to prepare a statement that can't be executed.

## Reproduction

```js
const mysql = require('mysql2');

const connection = mysql.createConnection({
  host: 'us-west.connect.psdb.cloud',
  user: '...',
  password: '...',
  database: 'mysqltest',
  ssl: { }

});

connection.query('drop table if exists test;')
connection.query('create table test(id int primary key);');
connection.query('insert into test values(1),(2),(3);');

// succeeds
var parameterCount = 65528;

// fails
// parameterCount = 65529; // fails and closes the connection
// parameterCount = 65536; // fails returns an error message

var questionMarks = Array(parameterCount).fill('?').join(',');
var values = [...Array(parameterCount).keys()];

connection.execute(
  `SELECT id FROM test WHERE id IN (${questionMarks})`,
  values,
  function(err, results, fields) {
    if (err) {
        console.log(err);
    } else {
        console.log(results);
    }
  }
);
```

**Comment by orware** [comment]
@bgrainger,

I've appreciated reading through the discussion posts you created over the weekend to share with us!

Reading through your post here there's a lot of good information and if the behavior is inconsistent in the way that you've already described above I'm just going to take this information straight to share as a potential Vitess issue on our end for additional review.

Based on what you've shared a fix may be in order, along with perhaps some potential tests to help ensure behavior is consistent when passing in that many parameters.

**Reply by bgrainger** [reply]
> I've appreciated reading through the discussion posts you created over the weekend to share with us!

Thanks!

Some background: I've written the most popular C#/.NET MySQL driver: https://github.com/mysql-net/MySqlConnector. When I saw a recent PlanetScale blog post (about HTTP/3) come up on HN recently, I was inspired to see if I could get my suite of integration tests passing with PlanetScale (using the MySQL protocol); the failing tests led to the creation of these bug reports.

If you ever open up the HTTP/3 API as an officially-supported feature, I might be interested in supporting it from the .NET client side of things...

**Reply by mattrobenolt** [reply]
@bgrainger 👋 send me an email and let's chat. matt@ or whatever address you find for me on GH.

**Reply by orware** [reply]
@bgrainger, I received word that the [following pull request](https://github.com/vitessio/vitess/pull/12614) for the Vitess project should address the issue you've shared with us here so it should be making its way into PlanetScale within a few weeks once it gets merged in.

Thanks again for the report!
